timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java

```diff
@@ -47,42 +46,75 @@ private static void checkValidTimestampRangeRequest(long numberToHandOut) {
         }
     }
 
+    /**
+     * Hand out the requested timestampRange, only increase the upper bound if strictly necessary. Do not block on the
+     * refresh of buffer if the existing buffer is already large enough.
+     *
+     * Modifies lastReturnedTimestamp and possibly upperLimit (but obtains a lock on this if it does).
+     *
+     * Possible race condition with refreshBuffer:
+     * targetTimestamp > upperLimit.get(), indicating that we need to increase the timestamp bound, but upperLimit
+     * gets increased before we obtain the lock in allocateEnoughTimestampsToHandOut. This is, however, not a problem
+     * because the check in upperLimit.increaseToAtLeast will use the fresh value of upperLimit.get() and therefore
+     * exit the method without performing an unnecessary cas.
+     *
+     * @param numberToHandOut number of timestamps
+     * @return
+     */
     public TimestampRange handOut(long numberToHandOut) {
         checkValidTimestampRangeRequest(numberToHandOut);
         long targetTimestamp;
         TimestampRange timestampRange;
 
-        synchronized (this) {
+        synchronized (issueTimestampLock) {
             /*
              * Under high concurrent load, this will be a hot critical section as clients request timestamps.
              * It is important to minimize contention as much as possible on this path.
              */
-            targetTimestamp = lastHandedOut() + numberToHandOut;
-            timestampRange = handOutTimestamp(targetTimestamp);
+            long lastHandedOut = lastReturnedTimestamp.get();
+            targetTimestamp = lastHandedOut + numberToHandOut;
+            timestampRange = TimestampRange.createInclusiveRange(lastHandedOut + 1, targetTimestamp);
+            if (targetTimestamp > upperLimit.get()) {
+                //this is the only situation where we actually must block and wait for CAS
+                allocateEnoughTimestampsToHandOut(targetTimestamp, ALLOCATION_BUFFER_SIZE);
```

Comment:
Yep, makes sense - the bound _must_ be at least targetTimestamp, but if we're doing a CAS let's bump it up a good bit more.